transforms/StickbreakingAngular/StickbreakingAngular_functions.stan

```diff
@@ -0,0 +1,46 @@
+vector stickbricking_angular_simplex_constrain_lp(vector y) {
+  int N = rows(y) + 1;
+  vector[N] x;
+  real log_phi, phi, u, s, c;
+  real s2_prod = 1;
+  real log_halfpi = log(pi()) - log2();
+  int rcounter = 2 * N - 3;
+  for (i in 1 : (N - 1)) {
+    u = log_inv_logit(y[i]);
```

Comment:
For `real`, it's conventional and just as efficient to declare and define at the same time.  The general consensus is that it's easier to read code if the declarations are as near the definitions as possible.

So that'd be

```stan
   real u = log_inv_logit(y[I]);
```

and so on.   With vectors, it can be more efficient to declare them once if there's going to be reallocation otherwise.

I would just go with whatever you find clearest here, but it's generally more efficient to vectorize the transforms, so that'd be

```stan
vector[N] log_phi = log_inv_logit(y) + (log(pi()) - log2());
```

and then replace uses of `log_phi` later with `log_phi[n]`.  The reason is that it reduces the size of the expression graph for autodiff.

Reply:
@bob-carpenter here's another example where I can completely eliminate the for loop for the log-simplex transform but with more vector allocations. Would this generally be more efficient in Stan?

```stan
vector stickbricking_angular_log_simplex_constrain_lp(vector y) {
  int N = rows(y) + 1;
  vector[N - 1] log_u = log_inv_logit(y);
  vector[N - 1] log_phi = log_u + log(pi()) - log2();
  vector[N - 1] phi = exp(log_phi);
  vector[N - 1] log_s = log(sin(phi));
  vector[N - 1] log_c = log(cos(phi));
  vector[N] log_s2_prod = cumulative_sum(append_row(0, 2 * log_s));
  vector[N] log_x = log_s2_prod + append_row(2 * log_c, 0);
  target += (N - 1) * log2();
  target += log_s2_prod;
  target += log_c;
  target += log_s;
  target += log_phi;
  target += log1m_exp(log_u);
  target += -log_x;
  return log_x;
}
```

Reply:
There's not a one-size-fits-all answer, I'm afraid.  It will generally depend on how big the vectors are, what other pressure there is on memory, and the amount of reduction in the expression graph and number of virtual function calls with vectorizations.

For example, this:

```stan
vector[N - 1] log_u = log_inv_logit(y);
```

will allocate the vector, but it will also reduce the number of virtual function calls to 1.  Whether that's a win will depend on compiler memory pressure, etc.  

I've rewritten the next line because arithmetic in C++ and hence Stan is evaluated left to right.  So it's better to combine scalars and then add to a vector.   

```stan
  vector[N - 1] log_phi = log_u + (log(pi()) - log2());
```

This should be faster because it simplifies the autodiff from a bunch of individual additions to vector addition, which simplifies the expression graph.  

The following

```stan
  vector[N - 1] phi = exp(log_phi);
```

is like the first expression, where the speedup or slowdown is going to depend on a lot of contingencies.

This

```stan
  vector[N] log_s2_prod = cumulative_sum(append_row(0, 2 * log_s));
```

is going to be a clear win over a loop because there's an efficient underlying gradient (I think---you'd really need to try it or look at the C++ code).  Now it'd probably be even faster to do the append_row on the outside. 

Then it depends on how they get used.  While creating the value might not be faster,

```stan
  target += -log_x;
```
is going to be a lot faster than doing it in a loop.

In general,

```stan
target += a;
target += b;
```

is faster than 

```stan
target += (a + b);
```

because under the hood, `target +=` just accumulates a list, which is fed into a final summation, which is more efficient than building up a binary tree of additions.